src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -1057,3 +1057,152 @@ cdef class DrawFuncs:
         self._close_path_func = func
         hb_draw_funcs_set_close_path_func(
             self._hb_drawfuncs, _close_path_func, <void*>user_data, NULL)
+
+cdef class HBLinks:
+    cdef hb_link_t* _hb_link
+
+    cdef set(self, hb_link_t* p_link):
+        self._hb_link = p_link
+
+    def add_links_value(self, links: List[Tuple[int, int, int]]):
+        cdef unsigned int num_links = len(links)
+        for i in range(num_links):
+            self._hb_link[i].position = links[i][0]
+            self._hb_link[i].width = links[i][1]
+            self._hb_link[i].objidx = links[i][2]
+
+cdef hb_link_t* create_links(unsigned int link_num,
+                             hb_object_t* obj_list,
+                             unsigned int length,
+                             hb_bool_t* success):
+    success[0] = 1
+    cdef hb_link_t* p = NULL
+    if link_num == 0:
+        return NULL
+    else:
+        p = <hb_link_t*>calloc(link_num, sizeof(hb_link_t))
+        if p == NULL:
+            free_allocated(obj_list, length)
+            # for virtual_links, need to free additional real_links which
+            # is created before virtual_links
+            if obj_list[length].real_links != NULL:
+                free(obj_list[length].real_links)
+            free(obj_list)
+
+            success[0] = 0
+            return NULL
+        else:
+            return p
+
+cdef void free_allocated(hb_object_t* obj_list, unsigned int length):
+    cdef unsigned int i
+    for i in range(length):
+        if obj_list[i].real_links != NULL:
+            free(obj_list[i].real_links)
+        if obj_list[i].virtual_links != NULL:
+            free(obj_list[i].virtual_links)
+
+# The whole table is represented as a Graph
+# The input graphnodes is a flat list of subtables
+# each Node(subtable) is represented by a tuple of 2 lists: real_links and virtual_links
```

Comment:
consider defining a Link namedtuple instead of using un-named tuples, so you can do link.objidx instead of link[2], which is much more readable. You can still accept a plain tuple from the caller and then convert it to your namedtuple, or you could expose these objects in the public python api and the caller needs to import them and use them to construct the graph to pass on to hb_repacker

```python
from typing import NamedTuple

class Link(NamedTuple):
    width: int
    position: int
    objidx: int
```

Reply:
typing.NamedTuple doesn't seem to work with python, see also:
https://github.com/harfbuzz/uharfbuzz/blob/main/src/uharfbuzz/_harfbuzz.pyx#L374

Reply:
Any reason collections.namedtuple can’t be used here then?

Reply:
collections.namedtuple doesn't support type hints.
Besides, it can not be defined as a subclass,  so I need to define it in both fonttools and cython.